ethcore/call-contract/src/call_contract.rs

```diff
@@ -18,12 +18,36 @@
 
 use bytes::Bytes;
 use ethereum_types::Address;
+use ethabi::{decode, ParamType, Token};
 use types::ids::BlockId;
 
+const ERROR_SELECTOR: [u8; 4] = [0x08, 0xc3, 0x79, 0xa0];
+
 /// Provides `call_contract` method
 pub trait CallContract {
```

Comment:
I have considerations regarding the entire trait, extracting revert messages really calls for introducing a proper error type here (at least a two-value enum).

I think it would also be way better to introduce an `Options` parameter (with a builder) instead of introducing another method:

```rust
pub enum CallError {
  Reverted(String),
  Other(String),
}

#[derive(derive_builder::Builder, Default, Clone)]
pub CallOptions {
  contract_address: Address,
  sender: Address,
  data: Bytes,
  value: U256,
  gas: U256,
  gas_price: U256,
}

impl CallOptions {
  /// Convenience method for creating the most common use case.
  pub fn new(contract: Address, data: Bytes) -> Self {
    Self::default().contract_address(contract).data(data)
  }
}

pub trait CallContract {
  /// Executes a transient call to a contract at given `BlockId`.
  /// 
  /// The constructed transaction must be executed on top of state at block `id`,
  /// any changes introduce by the call must be discarded.
  /// Returns:
  /// - A return data from the contract if the call was successful,
  /// - A `CallError::Reverted(msg)` error in case the call was reverted with an exception and message.
  /// - A `CallError::Other(msg)` in case the call did not succeed for other reasons.
  fn call_contract(&self, id: BlockId, call_options: CallOptions) -> Result<Bytes, CallError>;
}